core/lib/snapshots_applier/src/lib.rs

```diff
@@ -261,6 +261,41 @@ impl SnapshotsApplierTask {
         }
     }
 
+    /// Checks whether the snapshot recovery is already completed.
+    ///
+    /// Returns `None` if no snapshot recovery information is detected in the DB.
+    /// Returns `Some(true)` if the recovery is completed.
+    /// Returns `Some(false)` if the recovery is not completed.
+    pub async fn is_recovery_completed(
+        conn: &mut Connection<'_, Core>,
+        client: &dyn SnapshotsApplierMainNodeClient,
+    ) -> anyhow::Result<Option<bool>> {
+        let Some(applied_snapshot_status) = conn
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await?
+        else {
+            return Ok(None);
+        };
+        // If there are unprocessed storage logs chunks, the recovery is not complete.
+        if applied_snapshot_status.storage_logs_chunks_left_to_process() != 0 {
+            return Ok(Some(false));
+        }
+        // Currently, migrating tokens is the last step of the recovery.
+        // The number of tokens is not a part of the snapshot header, so we have to re-query the main node.
+        let added_tokens = conn.tokens_dal().get_all_l2_token_addresses().await?.len();
```

Comment:
Can't this number be greater than the number of recovered tokens if recovery was completed and the node has processed some blocks that introduce new tokens? Could make sense to use `TokensWeb3Dal::get_all_tokens()`, which is used in the `fetch_tokens` server handler.

Reply:
Great catch, thanks!